Reorg dirs, set up poetry

stratofax opened the pull request:
Some highlights:

- Makefile removed, poetry used exclusively
- The `.gitignore` file updated to allow poetry files to be included
- Created by following the steps in [Python: Using Poetry with an existing project in GitHub · ion-oset/developer-commons Wiki](https://github.com/ion-oset/developer-commons/wiki/Python:-Using-Poetry-with-an-existing-project-in-GitHub)

windoverwater commented on pyproject.toml:
```diff
@@ -0,0 +1,25 @@
+authors = ["Neil Johnson <neil@cadent.com>"]
+license = "GNU General Public License v2.0"
+readme = "README.md"
+packages = [{include = "vtp_rcv_demo"}]
```
Change this?

stratofax replied:
I believe we want to go with **vtp-spring-demo**?

windoverwater replied:
yes.  sure.

stratofax replied:
It turns out that I needed a name for the project that doesn't have hyphens, so I used the more descriptive "vtpserver" as the project name. This name works well for imports, and describes the function of the code much more accurately.

We can call the repo whatever you like, and use hyphens, too. So "vtp-spring-demo" or "vtp-sprint-demo" both work, but I'd actually argue for "vtp-api-server" because that's really what this is going to be.

windoverwater replied:
That is fine - I lean to more words such as vtpwebserver so to specify what type of server, but either works.

ion-oset replied:
Given that this a project called "sprint-demo" I would call it "demo-server". I wouldn't call it just "server" (there's all kinds of servers), or web-server (web servers are so common as to be the default). If it ultimately is intended to be a general VTP API then I'd call it "api-server" (though if that's the case it should end up in a different repository post-demo).

Other points to consider:

- Be aware that the PyPI rules for package naming are deliberately loose. `vtp-project`, `vtp.project`, and `vtp_project` are all the same as far as PyPI is concerned and `pip install` will work with any of them. Most people use dashes. There's also a case to be made for dots in the sense that it implies these are all part of a bigger project (`vtp.api-server`, an `vtp.test-suite` would obviously be in the same project, which is less clear if it's all dashes). I avoid underscores *because* dashes are not allowed in Python identifiers and there's no chance of confusion if they are used.
    - If I am wearing my (**informal!**) BDFL-on-naming hat, I'd call everything `vtp.{project-name-with-dashes}`.

- Please use *some* separator whatever you choose. wordsclusteredtogetherarehardtoread.

- The name can be changed up until the point it goes up on PyPI. I would strongly recommend *not* putting on PyPI until you are willing to commit to a naming scheme for projects as a whole. They cannot be changed easily.

- Last but not least: I recommend a read through [PEP 423 - Naming conventions and recipes related to packaging](https://peps.python.org/pep-0423/), specifically [Top-level namespace relates to code ownership](https://peps.python.org/pep-0423/#top-level-namespace-relates-to-code-ownership).

stratofax replied:
Here's what I found:

- [PEP 8 Package and Module Names](https://peps.python.org/pep-0008/#package-and-module-names] guidelines are succinct and quoted below. Recommended: underscores only in module names, if needed, but not in package names.
- Searching [PyPI](https://pypi.org/), I only found one package that includes the string "vtp": [vtp-elastic-apm 5.1.1](https://pypi.org/project/vtp-elastic-apm/). Note that "vtp" is not part of the namespace hierarchy, and after a brief perusal of this project's docs, I still can't tell what VTP stands for, either.

## Conclusion: "vtp" is a good top-level namespace for VoteTrackerPlus

The string "vtp" doesn't appear to be in use very much at all on PyPI, and is certainly not being used as a top-level namespace. Therefore, we could proceed ahead with using "vtp" as the top-level of the VoteTrackerPlus namespace, even if we never add any of the related repos to PyPI.

For this specific package, I favor the namespace: `vtp.api` as the shortest possible option, or `vtp.api_server` as a more descriptive option.

The repo name can be whatever we want, but I'm leaning toward something more functionally descriptive, like "VTP-API", or "VTP-API-Demo" if we want to manage expectations.

Here's the full text of the [PEP 8 Package and Module Names](https://peps.python.org/pep-0008/#package-and-module-names] guidelines:

> Modules should have short, all-lowercase names. Underscores can be used in the module name if it improves readability. Python packages should also have short, all-lowercase names, although the use of underscores is discouraged.

stratofax replied:
Based on today's meeting, we've articulated a namespace that works for VTP and applied it to this repo.

windoverwater commented:
No need to change this now, but I pretty strongly rely on pylint to catch all the static errors and warnings before looking at running code - you may want to add that as a dev dependency as well.

stratofax commented:
This should be the final update to this PR. I tested self-imports, and therefore the source code directory names and structure, and pyproject.toml paths, with my one test file, which runs correctly (and passes) when you execute `pytest` -- give it a go! 

Note how I removed all the dashes ("-") from the directories, because Python doesn't like dashes in import paths.

Also, added some more packages for dev and testing.

stratofax commented:
> No need to change this now, but I pretty strongly rely on pylint to catch all the static errors and warnings before looking at running code - you may want to add that as a dev dependency as well. ???

Agreed, and done!

stratofax commented:
@ion-oset I added you to this PR because this branch is focused on organizing a poetry project according to the "OSET way." Since you had more to do with defining what a good Python project structure looks like for OSET repos, I wanted you to take a look.

Thanks also to your suggestion on our Thursday call last week, I was able to create a `TestClient` test for the "Hello World" endpoint, which demonstrates both:

* That the `pyproject.toml` file and the corresponding directory structure correctly supports proper absolute path imports for this project
* That I can run `pytest` and test the API endpoints without having to set up a uvicorn server.

The next step is to build out the API but I'd like to do that on a new branch that's focused on that feature, based on a solid project structure, as defined in this PR. The hope is that we're close to closing this PR -- please let me know if you have any comments or changes.

Thanks!

ion-oset commented on pyproject.toml:
```diff
@@ -0,0 +1,28 @@
+packages = [{include = "vtpserver", from = "src"}]
+
+[tool.poetry.dependencies]
+python = "^3.11"
```
👉🏼 Important 👈🏼

Unless we have an explicit need for features in the latest Python release, we need to support earlier releases, definitely back to the most recent Ubuntu LTS (22.04) and ideally back at least a couple of years (if not all the way back to previous Ubuntu LTS (20.04)). I recommend this be no higher than 3.10, and ideally also consider supporting 3.9 which is going to still be in use for two more years (and past the date of the next national election).

This is not just a "nice-to-have": VTP is an unusual project in that the software meant to be run by regular users not just election administrators. While this is the server and not the VTP core tool (for which this is critical) if it's successful we should expect people to also look at the rest of the VTP projects. We can't require that people have to install their own Pythons beyond what's on the system to do it.

See:
- https://endoflife.date/python
- https://packages.ubuntu.com/jammy/python3
- https://packages.ubuntu.com/focal/python3

ion-oset commented on pyproject.toml:
```diff
@@ -0,0 +1,28 @@
+
+[tool.poetry.dependencies]
+python = "^3.11"
+requests = "^2.28.2"
```
Is this used in the sources or only in tests (in which case it should go in the dev dependencies)?

stratofax replied:
As noted elsewhere, I've dropped `requests` from the requirements.

ion-oset commented on pyproject.toml:
```diff
@@ -0,0 +1,28 @@
+uvicorn = "^0.20.0"
+httpx = "^0.23.3"
+
+[tool.poetry.group.dev.dependencies]
```
Alphabetize dependency lists.

stratofax replied:
Done.

ion-oset commented on src/vtpserver/api.py:
```diff
@@ -0,0 +1,17 @@
+async def get_vote_store_id() -> dict:
+    """Get a unique Vote Store ID for each client"""
+    vote_store_id = str(random.randrange(100000, 999999))
+    return {"VoteStoreID": vote_store_id}
```
👍🏼

ion-oset commented on tests/test_api.py:
```diff
@@ -0,0 +1,11 @@
+def test_read_main():
+    response = client.get("/")
+    assert response.status_code == 200
+    assert response.json() == {"message": "Hello World"}
```
👍🏼

stratofax replied:
This test file not only tests the API endpoint (with the status code -- nice!) but also demonstrates that our absolute import paths are working.

ion-oset reviewed (approved):
LGTM. See my notes on naming (both here and in the reply to the post about it).

Regarding the Makefile: taking it out for now is fine, but I am inclined to always have a Makefile. There are two reasons:

1. Makefiles are familiar to everyone regardless of language ecosystem. Projects in our space may end up being in more than one language.
2. They are effectively a quick menu of developer choices. They don't have to be complex. Things like 'make build', 'make test', and 'make serve' are what we are talking about. More advanced features can go elsewhere.

Regarding the READMEs: 👍🏼 on not having them all over. Better to have documentation eventually end up in `docs`.

stratofax commented on pyproject.toml:
```diff
@@ -0,0 +1,28 @@
+[tool.poetry]
+name = "vtpserver"
```
@windoverwater and I met today and discussed the namespace. Here's what we decided:

- The top-level namespace will be `vtp`
- The original VoteTrackerPlus package will be named `vtp.core`
- This package will be named `vtp.web.api`
- The web client package will be named `vtp.web.client`
- In conjunction with this new namespace, we've renamed this repo to `VTP-web-api`

@ion-oset  let us know what you think.

stratofax commented on pyproject.toml:
```diff
@@ -0,0 +1,28 @@
+requests = "^2.28.2"
+fastapi = "^0.92.0"
+uvicorn = "^0.20.0"
+httpx = "^0.23.3"
```
The `requests` package is no longer needed and I removed it; package names are sorted alphabetically, except for Python.

windoverwater commented:
Question - should the commits in this repo be signed?  The commits in the VoteTrackerPlus are required to be signed.   FWIIW I vote yes even though it is a hassle to set up, but once I set it I haven't touched it since them.

windoverwater commented:
I just checked - when I set up the repo I never enabled required signed commits for branches other than for the master branch - rats.  FWIIW it is now enabled on all branches in this repo.  It might be the case that the commits need to re-signed?  I don't know.  FWIIW I configure the signing of commits on a repository basis since I can push to different repos on different machines - I have been pushing signed commits to this one as well.  (They are supposed to show up on this page with a VERIFIED icon, which they are not.)

Regarding merge commits, my philosophy is that in the end they are bad (just my $0.02 cents) so yes I disabled them.  I think this is your repo so you get to decide.  Why I disable them on master is that I want master to have a linear history and github supports the 'restore the branch of a PR' with a button.  Since the branch can come back at any time, IMHO years from now the future will appreciate the linear history (with the power of reconstitution) more than the nonlinear history - forever.  Since version history is create once, read many (forever), I optimize for the future readers.  My IMHO though.

windoverwater commented:
p.s. I never disable merge commits on feature branches because my desire for linear history is only on the master branch - squash merging to the master removes all merge commits regardless of where they are.

stratofax commented:
> Merging is blocked: The base branch requires all commits to be signed.

For some reason, I can't view the verification status of my commits, even though I used an SSH public key for all of them.

@windoverwater you may have to turn on "vigilant mode" so we can audit the signatures: [Displaying verification statuses for all of your commits - GitHub Docs](https://docs.github.com/en/authentication/managing-commit-signature-verification/displaying-verification-statuses-for-all-of-your-commits)

windoverwater commented:
I don't think I have any commits in this branch.  On the other branch, my commits are showing up as verified but yours are not:

[https://github.com/TrustTheVote-Project/VTP-web-api/compare/master...feature/more-work-20230214](https://github.com/TrustTheVote-Project/VTP-web-api/compare/master...feature/more-work-20230214)

stratofax commented:
> can you verify that you really are signing the commits?

No, that's why I think you need to turn on "vigilant mode" so I can audit my signatures: [Displaying verification statuses for all of your commits - GitHub Docs](https://docs.github.com/en/authentication/managing-commit-signature-verification/displaying-verification-statuses-for-all-of-your-commits)

windoverwater commented:
I did turn it on (prior to the comment 41 minutes before this one :-)  ?

stratofax commented:
> I did turn it on (prior to the comment 41 minutes ago before this one :-) ?

And yet, I don't see the Verification status for my commits when I follow these steps: [Checking your commit and tag signature verification status - GitHub Docs](https://docs.github.com/en/authentication/troubleshooting-commit-signature-verification/checking-your-commit-and-tag-signature-verification-status)

windoverwater commented:
Correct.  I see mine but not yours.  Maybe try to push anyway.  If the push does not work, with signing really working, you could rebase the branch and that should in theory sign everything, and then it should be push-able.

Or I can disable signing, you could push, and I can re-enable it everywhere, which would make debugging it easier if it is still not working at that time.

?

stratofax commented:
> Correct. I see mine but not yours. Maybe try to push anyway. If the push does not work, with signing really working, you could rebase the branch and that should in theory sign everything, and then it should be push-able.
> 
> Or I can disable signing, you could push, and I can re-enable it everywhere, which would make debugging it easier if it is still not working at that time.

When you say "push" do you mean "merge?" Anyway, I can't merge: "Merging is blocked" and the **Squash and merge** button is disabled.

How about you do the merge, and disable signing if needed? 

I can't tell if I'm signing anything, because I can't see any verification status info for any of my commits.
